core/src/main/scala/org/apache/spark/executor/ExecutorMetricsPoller.scala

```diff
@@ -0,0 +1,202 @@
+import org.apache.spark.util.{ThreadUtils, Utils}
+
+/**
+ * :: DeveloperApi ::
+ * A class that polls executor metrics, and tracks their peaks per task and per stage.
+ * Each executor keeps an instance of this class.
+ * The poll method polls the executor metrics, and is either run in its own thread or
+ * called by the executor's heartbeater thread, depending on configuration.
+ * The class keeps two ConcurrentHashMaps that are accessed (via its methods) by the
+ * executor's task runner threads concurrently with the polling thread. One thread may
+ * update one of these maps while another reads it, so the reading thread may not get
+ * the latest metrics, but this is ok.
+ *
+ * @param memoryManager the memory manager used by the executor.
+ * @param pollingInterval the polling interval in milliseconds.
+ */
+@DeveloperApi
+private[spark] class ExecutorMetricsPoller(
+    memoryManager: MemoryManager,
+    pollingInterval: Long)
+  extends Logging {
+
+  type StageKey = (Int, Int)
+  // tuple for Task Count and Metric Peaks
+  type TCMP = (AtomicLong, AtomicLongArray)
+
+  // Map of (stageId, stageAttemptId) to (count of running tasks, executor metric peaks)
+  private val stageTCMP = new ConcurrentHashMap[StageKey, TCMP]
+
+  // Map of taskId to executor metric peaks
+  private val taskMetricPeaks = new ConcurrentHashMap[Long, AtomicLongArray]
+
+  private val poller =
+    if (pollingInterval > 0) {
+      ThreadUtils.newDaemonSingleThreadScheduledExecutor("executor-metrics-poller")
+    } else {
+      null
+    }
+
+  /**
+   * Function to poll executor metrics.
+   * On start, if pollingInterval is positive, this is scheduled to run at that interval.
+   * Otherwise, this is called by the reportHeartBeat function defined in Executor and passed
+   * to its Heartbeater.
+   */
+  def poll(): Unit = {
+    // Note: Task runner threads may update stageTCMP or read from taskMetricPeaks concurrently
+    // with this function via calls to methods of this class.
+
+    // get the latest values for the metrics
+    val latestMetrics = ExecutorMetrics.getCurrentMetrics(memoryManager)
+
+    def compareAndUpdate(current: Long, latest: Long): Long =
+      if (latest > current) latest else current
+
+    def updatePeaks(metrics: AtomicLongArray): Unit = {
+      (0 until metrics.length).foreach { i =>
+        metrics.getAndAccumulate(i, latestMetrics(i), compareAndUpdate)
+      }
+    }
+
+    // for each active stage, update the peaks
+    stageTCMP.forEachValue(LONG_MAX_VALUE, v => updatePeaks(v._2))
+
+    // for each running task, update the peaks
+    taskMetricPeaks.forEachValue(LONG_MAX_VALUE, updatePeaks)
+  }
+
+  /** Starts the polling thread. */
+  def start(): Unit = {
+    if (poller != null) {
+      val pollingTask = new Runnable() {
+        override def run(): Unit = Utils.logUncaughtExceptions(poll())
+      }
+      poller.scheduleAtFixedRate(pollingTask, 0L, pollingInterval, TimeUnit.MILLISECONDS)
```

Comment:
minor: with Scala 2.12 and java8, I think you don't even need to make this anonymous `Runnable` anymore:

https://www.scala-lang.org/news/2.12.0/#lambda-syntax-for-sam-types

https://github.com/apache/spark/pull/24241

Reply:
Changed to use lambda.